From 1e4ed1d4abbcfbf02ef80e000ce8ebd841dcf667 Mon Sep 17 00:00:00 2001
Message-ID: <1e4ed1d4abbcfbf02ef80e000ce8ebd841dcf667.1783491438.git.robux4@ycbcr.xyz>
From: =?UTF-8?q?Hugo=20Beauz=C3=A9e-Luyssen?= <hugo@beauzee.fr>
Date: Thu, 28 Mar 2019 15:23:48 +0100
Subject: [PATCH 01/20] compat: Workaround sendmsg bug on android

This only happens on 64bits builds, see compat/sendmsg.c comments
Message-Id: <b75b490465c538e50780039dc796724974033db0.1593695247.git.thomas@gllm.fr>
Message-Id: <87030933b1aae776c8446cf765589f8ac8b5ba7c.1599040437.git.nicolas@videolabs.io>
Message-Id: <4b80ba0cb27cadee99cf18b8f13fd7d16f19066b.1601374253.git.nicolas@videolabs.io>
Message-Id: <7911214e6bde9e4c74765cea26c5944e8e78ca98.1602659942.git.nicolas@videolabs.io>
Message-Id: <155aceebce778fcb6fb6877de5bb1cf066f99c6f.1605192544.git.nicolas@videolabs.io>
Message-Id: <fcaa38ed543fd88a73fa3c6766a8f86b8b053fb6.1605701951.git.nicolas@videolabs.io>
Message-Id: <7505e758ab3f173e530063cea6a221862fb83ab2.1607001652.git.nicolas@videolabs.io>
Message-Id: <725e37106e83190ea20042aa9dc01c1c7646b37d.1628587500.git.thomas@gllm.fr>
Message-Id: <c39c6a221ade0e2440f3a10d57b3ecf33e982311.1657539375.git.thomas@gllm.fr>
Message-Id: <8db168c0fbd0849e5bbc7b852b4cf9e61a5866b9.1662737834.git.dev.asenat@posteo.net>
Message-Id: <9c6326a62110dc03fa3c8a561747ff73ce2e7fcc.1664884159.git.dev.asenat@posteo.net>
Message-Id: <1c60b920e01d183ec81d9c862015e336b11b7816.1664886557.git.dev.asenat@posteo.net>
Message-Id: <e0f1ebfe2353a812b00011cddea923b56fa79c4a.1738244316.git.dcn.mcnamara@gmail.com>
---
 compat/sendmsg.c      | 24 ++++++++++++++++++++++++
 configure.ac          |  3 +++
 include/vlc_network.h |  6 ++++++
 3 files changed, 33 insertions(+)

diff --git a/compat/sendmsg.c b/compat/sendmsg.c
index 36470623e3..760b529b84 100644
--- a/compat/sendmsg.c
+++ b/compat/sendmsg.c
@@ -169,6 +169,30 @@ ssize_t sendmsg(int fd, const struct msghdr *msg, int flags)
     free(data);
     return res;
 }
+#elif defined(__ANDROID__) && defined(__aarch64__)
+
+#undef sendmsg
+
+#include <sys/types.h>
+#include <sys/socket.h>
+
+/**
+ * Since we bumped the NDK version from 14 to 18, some devices (at least up to
+ * Android 6) are returning errors on 4 bytes, even though ssize_t is actually
+ * 8 bytes. This causes `value < 0` checks to yield false, and consider the value
+ * as a non error.
+ * As the patch lies in either the NDK or the Android kernel, or the device libc
+ * we can only work around it. If errno is not 0 & we receive -1, on 32bits or
+ * 64bits, we assume an error was returned.
+ */
+ssize_t vlc_sendmsg(int fd, const struct msghdr *msg, int flags)
+{
+    errno = 0;
+    ssize_t ret = sendmsg(fd, msg, flags);
+    if ((ret < 0 || ret == 0xFFFFFFFF) && errno != 0)
+        return -1;
+    return ret;
+}
 
 #else
 #error sendmsg not implemented on your platform!
diff --git a/configure.ac b/configure.ac
index 654b47e05a..214129c9ea 100644
--- a/configure.ac
+++ b/configure.ac
@@ -458,6 +458,9 @@ AS_IF([test "$SYS" = linux],[
     ],[
       HAVE_ANDROID="1"
       AC_MSG_RESULT([yes])
+      AS_IF([test "${host_cpu}" = "aarch64"], [
+          AC_LIBOBJ([sendmsg])
+      ])
     ],[
       AC_MSG_RESULT([no])
     ])
diff --git a/include/vlc_network.h b/include/vlc_network.h
index e04124e815..758d92f756 100644
--- a/include/vlc_network.h
+++ b/include/vlc_network.h
@@ -305,6 +305,12 @@ static inline void net_SetPort (struct sockaddr *addr, uint16_t port)
 
 VLC_API char *vlc_getProxyUrl(const char *);
 
+#if defined(__ANDROID__) && defined(__aarch64__)
+struct msghdr;
+ssize_t vlc_sendmsg(int fd, const struct msghdr *msg, int flags);
+#define sendmsg(a, b, c) vlc_sendmsg(a,b,c)
+#endif
+
 # ifdef __cplusplus
 }
 # endif
-- 
2.51.0

